airflow/api_connexion/endpoints/task_instance_endpoint.py

```diff
@@ -14,23 +14,110 @@
+
+
+@provide_session
+def get_task_instance(dag_id: str, dag_run_id: str, task_id: str, session=None):
+    query = (
+        session.query(TI)
+        .filter(TI.dag_id == dag_id)
+        .join(DR, and_(TI.dag_id == DR.dag_id, TI.execution_date == DR.execution_date))
+        .filter(DR.run_id == dag_run_id)
+        .filter(TI.task_id == task_id)
+    )
+
+    task_instance = query.one_or_none()
+
+    if task_instance is None:
+        raise NotFound("Task instance not found")
+
+    return task_instance_schema.dump(task_instance)
 
-def get_task_instance():
-    """
-    Get a task instance
-    """
-    raise NotImplementedError("Not implemented yet.")
 
+def _apply_array_filter(query, key, values):
+    if values is not None:
+        query = query.filter(or_(*[key == v for v in values]))
+    return query
 
-def get_task_instances():
+
+def _apply_range_filter(query, key, value_range: Tuple[Any, Any]):
+    gte_value, lte_value = value_range
+    if gte_value is not None:
+        query = query.filter(key >= gte_value)
+    if lte_value is not None:
+        query = query.filter(key <= lte_value)
+    return query
+
+
+@format_parameters(
+    {
+        'start_date_gte': format_datetime,
+        'start_date_lte': format_datetime,
+        'execution_date_gte': format_datetime,
+        'execution_date_lte': format_datetime,
+        'end_date_gte': format_datetime,
+        'end_date_lte': format_datetime,
+    }
+)
+@provide_session
+def get_task_instances(
+    limit: int,
+    dag_id: Optional[str] = None,
+    dag_run_id: Optional[str] = None,
+    execution_date_gte: Optional[str] = None,
+    execution_date_lte: Optional[str] = None,
+    start_date_gte: Optional[str] = None,
+    start_date_lte: Optional[str] = None,
+    end_date_gte: Optional[str] = None,
+    end_date_lte: Optional[str] = None,
+    duration_gte: Optional[float] = None,
+    duration_lte: Optional[float] = None,
+    state: Optional[str] = None,
+    pool: Optional[List[str]] = None,
+    queue: Optional[List[str]] = None,
+    offset: Optional[int] = None,
+    session=None,
+):
     """
-    Get list of task instances of DAG.
+    Get list of a task instances
     """
-    raise NotImplementedError("Not implemented yet.")
+    query = session.query(TI)
+
+    if dag_id is not None:
+        query = query.filter(TI.dag_id == dag_id)
+    if dag_run_id is not None:
+        query = query.join(DR, and_(TI.dag_id == DR.dag_id, TI.execution_date == DR.execution_date))
+        query = query.filter(DR.run_id == dag_run_id)
+
+    query = _apply_range_filter(
+        query, key=DR.execution_date, value_range=(execution_date_gte, execution_date_lte)
+    )
+    query = _apply_range_filter(query, key=DR.start_date, value_range=(start_date_gte, start_date_lte))
+    query = _apply_range_filter(query, key=DR.end_date, value_range=(end_date_gte, end_date_lte))
+    query = _apply_range_filter(query, key=DR.end_date, value_range=(end_date_gte, end_date_lte))
```

Comment:
Looks like the preceding two lines are duplicates

Reply:
Good point. Itt should be execution_date.